src/TipTapEditor/DraggableBlock/DraggableBlockNode.ts

```diff
@@ -0,0 +1,112 @@
+import { Node } from '@tiptap/core';
+import { Fragment, Slice } from '@tiptap/pm/model';
+import { NodeSelection, TextSelection } from '@tiptap/pm/state';
+import { ReplaceStep } from '@tiptap/pm/transform';
+import { mergeAttributes, ReactNodeViewRenderer } from '@tiptap/react';
+
+import { DraggableBlock } from './DraggableBlock';
+
+declare module '@tiptap/core' {
+  interface Commands<ReturnType> {
+    splitDraggableBlock: {
+      splitDraggableBlock: () => ReturnType;
+    };
+  }
+}
+
+export const DraggableBlockNode = Node.create({
+  name: 'draggableBlock',
+
+  content: 'block',
+  defining: true,
+  draggable: true,
+  selectable: true,
+
+  parseHTML() {
+    return [
+      {
+        tag: 'div[data-type="draggable-block"]',
+      },
+    ];
+  },
+
+  renderHTML({ HTMLAttributes }) {
+    return ['div', mergeAttributes(HTMLAttributes, { 'data-type': 'draggable-block' }), 0];
+  },
+
+  addNodeView() {
+    return ReactNodeViewRenderer(DraggableBlock);
+  },
+
+  addCommands() {
+    return {
+      ...this.parent?.(),
+      splitDraggableBlock:
+        () =>
+        ({ tr, state, dispatch }) => {
+          const { selection } = state;
+          const { $from, $to } = selection;
+
+          if (('node' in selection && (selection as NodeSelection).node.isBlock) || $from.depth < 2) {
+            return false;
+          }
+
+          let draggableBlockDepth = $from.depth;
+          while (draggableBlockDepth > 0 && $from.node(draggableBlockDepth).type.name !== this.type.name) {
+            draggableBlockDepth -= 1;
+          }
+          // If selection does not belong to a draggable block, this command does not run
+          if (draggableBlockDepth === 0) {
+            return false;
+          }
+
+          // Create a copy of the block, sliced to before the selection
+          let preNode = $from.parent.type.create(
+            $from.parent.attrs,
+            $from.parent.slice(0, $from.parentOffset).content,
+            $from.parent.marks,
+          );
+          for (let i = $from.depth - 1; i >= draggableBlockDepth; i--) {
+            const node = $from.node(i);
+            preNode = node.type.create(node.attrs, preNode, node.marks);
+          }
+
+          // Create a new draggable block, containing text after the selection
+          const postNode = this.type.create(
+            null,
+            $to.parent.type.create($to.parent.attrs, $to.parent.slice($to.parentOffset).content, $to.parent.marks),
+          );
+
+          if (dispatch) {
```

Comment:
Maybe you can push this check to the top.

```js
if(!dispatch) {
  return
}
```

Reply:
No, we actually want to return true even if dispatch is null or undefined. Prosemirror's commands can do dry-runs by passing a null dispatch. In this case, the command should still return the boolean indicating whether the command can run. From what I have seen from TipTap's repository, the usual way to do it it to have some checks first, that can make the command return false, then the transformations in an `if (dispatch) {` and finally `return true;`. Here is an example for the updateAttributes command: https://github.com/ueberdosis/tiptap/blob/b0198eb14b98db5ca691bd9bfe698ffaddbc4ded/packages/core/src/commands/updateAttributes.ts#L43